rpcs3/rpcs3qt/update_manager.cpp

```diff
@@ -0,0 +1,518 @@
+
+		if (m_progress_dialog)
+		{
+			m_progress_dialog->close();
+			m_progress_dialog = nullptr;
+		}
+	}
+
+	return true;
+}
+
+bool update_manager::handle_json(const QByteArray& data, bool automatic)
+{
+	const QJsonObject json_data = QJsonDocument::fromJson(data).object();
+	int return_code             = json_data["return_code"].toInt(-3);
+
+	bool hash_found = true;
+
+	if (return_code < 0)
+	{
+		std::string error_message;
+		switch (return_code)
+		{
+		case -1: error_message = "Hash not found"; break;
+		case -2: error_message = "Server Error - Maintenance Mode"; break;
+		case -3: error_message = "Server Error - Return code not found"; break;
+		default: error_message = "Server Error - Unknown Error"; break;
+		}
+
+		LOG_ERROR(GENERAL, "[Auto-updater] error: %s return code: %d", error_message, return_code);
+
+		// If a user clicks "Check for Updates" with a custom build ask him if he's sure he wants to update to latest version
+		if (!automatic && return_code == -1)
+		{
+			if (QMessageBox::question(m_progress_dialog, tr("Custom Build"), tr("You're currently using a custom build.\nDo you want to update to the latest official rpcs3 version?"),
+			        QMessageBox::Yes | QMessageBox::No) == QMessageBox::No)
+			{
+				if (m_progress_dialog)
+				{
+					m_progress_dialog->close();
+					m_progress_dialog = nullptr;
+				}
+				return true;
+			}
+
+			hash_found = false;
+		}
+		else
+		{
+			return false;
+		}
+	}
+
+	const auto& latest = json_data["latest_build"];
+	if (!latest.isObject())
+	{
+		LOG_ERROR(GENERAL, "[Auto-updater] Json doesn't contain latest_build section");
+		return false;
+	}
+
+	QString os;
+
+#ifdef _WIN32
+	os = "windows";
+#elif defined(__linux__)
+	os = "linux";
+#else
+	LOG_ERROR(GENERAL, "[Auto-updater] Your os isn't currently supported by the auto-updater");
+	return false;
+#endif
+
+	if (!latest[os].isObject() || !latest[os]["download"].isString() || !latest["version"].isString() || !latest[os]["size"].isDouble() || !latest[os]["checksum"].isString())
+	{
+		LOG_ERROR(GENERAL, "[Auto-updater] Some information seems unavailable", os.toStdString());
+		return false;
+	}
+
+	std::string remote_version    = latest["version"].toString().toStdString();
+	const auto version_and_commit = fmt::split(remote_version, {"-"});
+	if (version_and_commit.size() != 2)
+	{
+		LOG_ERROR(GENERAL, "[Auto-updater] Version returned seems invalid: %s", remote_version);
+		return false;
+	}
+
+	u32 cur_commit    = std::stoul(rpcs3::get_commit_and_hash().first);
+	u32 remote_commit = std::stoul(version_and_commit[1]);
+
+	m_expected_hash = latest[os]["checksum"].toString().toStdString();
+	m_expected_size = latest[os]["size"].toInt();
+
+	if (hash_found && cur_commit >= remote_commit)
+		return false;
+
+	if (hash_found)
+	{
+		const QString to_show = tr("A new version of RPCS3 is available!\nLocal version: ") + QString::fromStdString(rpcs3::get_commit_and_hash().first) + tr("\nRemote version: ") +
```

Comment:
This is awful :)
You should always use a formatted QString when mixing values with translated strings.
Something like:
`tr("some text %0/n some other text %1 %2").args(string0, string1, string2)`

Reply:
I couldn't remember QT weird syntax and just went with a way I knew worked.

Reply:
This would accumulate a lot of issues for localization in the long run